source/opt/folding_rules.cpp

```diff
@@ -1607,27 +1607,26 @@ bool CompositeConstructFeedingExtract(
 }
 
 // Walks the indexes chain from |start| to |end| of an OpCompositeInsert or
-// OpCompositeExtract instruction, and returns the type of the final element
-// being accessed.
-const analysis::Type* GetElementType(uint32_t type_id,
-                                     Instruction::iterator start,
-                                     Instruction::iterator end,
-                                     const analysis::TypeManager* type_mgr) {
-  const analysis::Type* type = type_mgr->GetType(type_id);
+// OpCompositeExtract instruction, and returns the type id of the final element
+// being accessed. Returns 0 if a valid type could not be found.
+uint32_t GetElementType(uint32_t type_id, Instruction::iterator start,
```

Comment:
returning a valid ID (0) bugs me a bit.
Should this return a `std::optional` or something else we know cannot be valid?

Reply:
`0` is not a reserved id in spir-v. So the common idiom on spir-v tools is to return 0 the same way that nullptr is used.

@alan-baker Do you know where in the spec it says that? I looked quickly, but could not find it.

Reply:
Under [section 2.3](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_physical_layout_of_a_spir_v_module_and_instruction), Physical Layout it says:
> Bound; where all `<id>` in this module are guaranteed to satisfy
>`0 < id < Bound`

Reply:
ah right, thanks!